### PR TITLE
[candi](openstack) Add ServerGroup support for master nodes

### LINKED ISSUE #5958: ServerGroup into the Openstack
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Use case. Why is this important?

We need to create ServerGroup in Openstack, during the bootstrap, when masters are creating. This is necessary to enable anti-affinity, cause placing all masters on the same hyper becomes a problem.

### Proposed Solution

Terraform allows to register Server Group by scheduler_hints and link to instances.

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by artbin**
## Description

Add ServerGroup support for master nodes.

## Why do we need it, and what problem does it solve?

Close #5958.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: candi
type: feature
summary: Add OpenStack server group support for master nodes.
impact_level: default
```

### CODE REVIEW

**Comment by artbin** [comment]
Manual testing
==============

`OpenStackClusterConfiguration` without `serverGroup`:
```
apiVersion: deckhouse.io/v1
kind: OpenStackClusterConfiguration
layout: Standard
masterNodeGroup:
  instanceClass:
    additionalSecurityGroups:
    - node-group-static-frontend
    etcdDiskSizeGb: 10
    flavorName: m1.large
    imageName: ubuntu-22-04-cloud-amd64
  replicas: 1
  volumeTypeMap:
    nova: ceph-ssd
...
...
```

`OpenStackClusterConfiguration` with `serverGroup.name`:
```
apiVersion: deckhouse.io/v1
kind: OpenStackClusterConfiguration
layout: Standard
masterNodeGroup:
  instanceClass:
    additionalSecurityGroups:
    - node-group-static-frontend
    etcdDiskSizeGb: 10
    flavorName: m1.large
    imageName: ubuntu-22-04-cloud-amd64
  replicas: 1
  volumeTypeMap:
    nova: ceph-ssd
  serverGroup:
    name: master
...
...
```

`dhctl converge` output:
```
│ │ │ An execution plan has been generated and is shown below.
│ │ │ Resource actions are indicated with the following symbols:
│ │ │   + create
│ │ │ -/+ destroy and then create replacement
│ │ │ 
│ │ │ Terraform will perform the following actions:
│ │ │ 
│ │ │   # module.kubernetes_data.openstack_compute_volume_attach_v2.kubernetes_data must be replaced
│ │ │ -/+ resource "openstack_compute_volume_attach_v2" "kubernetes_data" {
│ │ │       ~ device      = "/dev/vdb" -> (known after apply)
│ │ │       ~ id          = "e21f77ab-441f-4759-801d-4b175dad3c04/66ce51b5-c733-45bc-9937-0bb315c3d910" -> (known after apply)
│ │ │       ~ instance_id = "e21f77ab-441f-4759-801d-4b175dad3c04" -> (known after apply) # forces replacement
│ │ │       ~ region      = "HetznerFinland" -> (known after apply)
│ │ │         volume_id   = "66ce51b5-c733-45bc-9937-0bb315c3d910"
│ │ │     }
│ │ │ 
│ │ │   # module.master.openstack_compute_floatingip_associate_v2.master[0] must be replaced
│ │ │ -/+ resource "openstack_compute_floatingip_associate_v2" "master" {
│ │ │         floating_ip           = "95.217.123.53"
│ │ │       ~ id                    = "95.217.123.53/e21f77ab-441f-4759-801d-4b175dad3c04/" -> (known after apply)
│ │ │       ~ instance_id           = "e21f77ab-441f-4759-801d-4b175dad3c04" -> (known after apply) # forces replacement
│ │ │       ~ region                = "HetznerFinland" -> (known after apply)
│ │ │         wait_until_associated = true
│ │ │     }
│ │ │ 
│ │ │   # module.master.openstack_compute_instance_v2.master must be replaced
│ │ │ -/+ resource "openstack_compute_instance_v2" "master" {
│ │ │       ~ access_ip_v4        = "192.168.199.190" -> (known after apply)
│ │ │       + access_ip_v6        = (known after apply)
│ │ │       ~ all_metadata        = {} -> (known after apply)
│ │ │       ~ all_tags            = [] -> (known after apply)
│ │ │         availability_zone   = "nova"
│ │ │         config_drive        = false
│ │ │       ~ flavor_id           = "69a1095b-ceec-404f-8545-800391cdbce1" -> (known after apply)
│ │ │         flavor_name         = "m1.large"
│ │ │         force_delete        = false
│ │ │       ~ id                  = "e21f77ab-441f-4759-801d-4b175dad3c04" -> (known after apply)
│ │ │       ~ image_id            = "6b6eff83-c807-4f05-90a4-c67eab2462b6" -> (known after apply)
│ │ │         image_name          = "ubuntu-22-04-cloud-amd64"
│ │ │         key_pair            = "node-group-static3"
│ │ │         name                = "node-group-static3-master-0"
│ │ │         power_state         = "active"
│ │ │       ~ region              = "HetznerFinland" -> (known after apply)
│ │ │       ~ security_groups     = [
│ │ │           - "node-group-static3",
│ │ │           - "node-group-static3-frontend",
│ │ │         ] -> (known after apply)
│ │ │         stop_before_destroy = false
│ │ │       - tags                = [] -> null
│ │ │       + user_data           = "07c7dbe45e392db72514060d4954540e7ccb8129"
│ │ │ 
│ │ │       ~ network {
│ │ │             access_network = false
│ │ │           ~ fixed_ip_v4    = "192.168.199.190" -> (known after apply)
│ │ │           + fixed_ip_v6    = (known after apply)
│ │ │           + floating_ip    = (known after apply)
│ │ │           ~ mac            = "fa:16:3e:16:ff:2d" -> (known after apply)
│ │ │           ~ name           = "node-group-static3" -> (known after apply)
│ │ │             port           = "376c3de1-9a7a-4633-8277-fdc9dd618aea"
│ │ │           ~ uuid           = "9d9d4176-bb47-409b-adde-22aa0d19e9c6" -> (known after apply)
│ │ │         }
│ │ │ 
│ │ │       + scheduler_hints { # forces replacement
│ │ │           + different_host = []
│ │ │           + group          = (known after apply)
│ │ │           + query          = []
│ │ │           + same_host      = []
│ │ │         }
│ │ │     }
│ │ │ 
│ │ │   # module.master.openstack_compute_servergroup_v2.master[0] will be created
│ │ │   + resource "openstack_compute_servergroup_v2" "master" {
│ │ │       + id       = (known after apply)
│ │ │       + members  = (known after apply)
│ │ │       + name     = "master"
│ │ │       + policies = [
│ │ │           + "anti-affinity",
│ │ │         ]
│ │ │       + region   = (known after apply)
│ │ │     }
│ │ │ 
│ │ │ Plan: 4 to add, 0 to change, 3 to destroy.
│ │ │ 
│ │ │ Changes to Outputs:
│ │ │   ~ kubernetes_data_device_path = "/dev/vdb" -> (known after apply)
│ │ │   ~ node_internal_ip_address    = "192.168.199.190" -> (known after apply)
│ │ │ Terraform runner "master-node" process exited.
```

`OpenStackClusterConfiguration` with `serverGroup.id`:
```
apiVersion: deckhouse.io/v1
kind: OpenStackClusterConfiguration
layout: Standard
masterNodeGroup:
  instanceClass:
    additionalSecurityGroups:
    - node-group-static-frontend
    etcdDiskSizeGb: 10
    flavorName: m1.large
    imageName: ubuntu-22-04-cloud-amd64
  replicas: 1
  volumeTypeMap:
    nova: ceph-ssd
  serverGroup:
    id: aaf6102d-7aaf-4ced-8937-2ea1144373a7
...
...
```

`dhctl converge` output:
```
│ │ │ An execution plan has been generated and is shown below.
│ │ │ Resource actions are indicated with the following symbols:
│ │ │ -/+ destroy and then create replacement
│ │ │ 
│ │ │ Terraform will perform the following actions:
│ │ │ 
│ │ │   # module.kubernetes_data.openstack_compute_volume_attach_v2.kubernetes_data must be replaced
│ │ │ -/+ resource "openstack_compute_volume_attach_v2" "kubernetes_data" {
│ │ │       ~ device      = "/dev/vdb" -> (known after apply)
│ │ │       ~ id          = "e21f77ab-441f-4759-801d-4b175dad3c04/66ce51b5-c733-45bc-9937-0bb315c3d910" -> (known after apply)
│ │ │       ~ instance_id = "e21f77ab-441f-4759-801d-4b175dad3c04" -> (known after apply) # forces replacement
│ │ │       ~ region      = "HetznerFinland" -> (known after apply)
│ │ │         volume_id   = "66ce51b5-c733-45bc-9937-0bb315c3d910"
│ │ │     }
│ │ │ 
│ │ │   # module.master.openstack_compute_floatingip_associate_v2.master[0] must be replaced
│ │ │ -/+ resource "openstack_compute_floatingip_associate_v2" "master" {
│ │ │         floating_ip           = "95.217.123.53"
│ │ │       ~ id                    = "95.217.123.53/e21f77ab-441f-4759-801d-4b175dad3c04/" -> (known after apply)
│ │ │       ~ instance_id           = "e21f77ab-441f-4759-801d-4b175dad3c04" -> (known after apply) # forces replacement
│ │ │       ~ region                = "HetznerFinland" -> (known after apply)
│ │ │         wait_until_associated = true
│ │ │     }
│ │ │ 
│ │ │   # module.master.openstack_compute_instance_v2.master must be replaced
│ │ │ -/+ resource "openstack_compute_instance_v2" "master" {
│ │ │       ~ access_ip_v4        = "192.168.199.190" -> (known after apply)
│ │ │       + access_ip_v6        = (known after apply)
│ │ │       ~ all_metadata        = {} -> (known after apply)
│ │ │       ~ all_tags            = [] -> (known after apply)
│ │ │         availability_zone   = "nova"
│ │ │         config_drive        = false
│ │ │       ~ flavor_id           = "69a1095b-ceec-404f-8545-800391cdbce1" -> (known after apply)
│ │ │         flavor_name         = "m1.large"
│ │ │         force_delete        = false
│ │ │       ~ id                  = "e21f77ab-441f-4759-801d-4b175dad3c04" -> (known after apply)
│ │ │       ~ image_id            = "6b6eff83-c807-4f05-90a4-c67eab2462b6" -> (known after apply)
│ │ │         image_name          = "ubuntu-22-04-cloud-amd64"
│ │ │         key_pair            = "node-group-static3"
│ │ │         name                = "node-group-static3-master-0"
│ │ │         power_state         = "active"
│ │ │       ~ region              = "HetznerFinland" -> (known after apply)
│ │ │       ~ security_groups     = [
│ │ │           - "node-group-static3",
│ │ │           - "node-group-static3-frontend",
│ │ │         ] -> (known after apply)
│ │ │         stop_before_destroy = false
│ │ │       - tags                = [] -> null
│ │ │       + user_data           = "07c7dbe45e392db72514060d4954540e7ccb8129"
│ │ │ 
│ │ │       ~ network {
│ │ │             access_network = false
│ │ │           ~ fixed_ip_v4    = "192.168.199.190" -> (known after apply)
│ │ │           + fixed_ip_v6    = (known after apply)
│ │ │           + floating_ip    = (known after apply)
│ │ │           ~ mac            = "fa:16:3e:16:ff:2d" -> (known after apply)
│ │ │           ~ name           = "node-group-static3" -> (known after apply)
│ │ │             port           = "376c3de1-9a7a-4633-8277-fdc9dd618aea"
│ │ │           ~ uuid           = "9d9d4176-bb47-409b-adde-22aa0d19e9c6" -> (known after apply)
│ │ │         }
│ │ │ 
│ │ │       + scheduler_hints { # forces replacement
│ │ │           + different_host = []
│ │ │           + group          = "aaf6102d-7aaf-4ced-8937-2ea1144373a7"
│ │ │           + query          = []
│ │ │           + same_host      = []
│ │ │         }
│ │ │     }
│ │ │ 
│ │ │ Plan: 3 to add, 0 to change, 3 to destroy.
│ │ │ 
│ │ │ Changes to Outputs:
│ │ │   ~ kubernetes_data_device_path = "/dev/vdb" -> (known after apply)
│ │ │   ~ node_internal_ip_address    = "192.168.199.190" -> (known after apply)
│ │ │ Terraform runner "master-node" process exited.
```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @artbin. I've started the workflow for label 'e2e/run/openstack'...

**Review by aleksey-su — CHANGES_REQUESTED** [review]
Russian documentation is needed

**Comment by z9r5** [comment]
Documentation is in WIP state. Don't merge.

**Review by RomanenkoDenys — CHANGES_REQUESTED** [review]
Looks good, one simple comment

**Review by name212 — APPROVED** [review]

File: ee/candi/cloud-providers/openstack/openapi/cluster_configuration.yaml [hunk 1]
```diff
@@ -197,6 +197,41 @@ apiVersions:
...
+                    type: string
+                    description: |
+                      The id of the existing `ServerGroup` object managed outside of Deckhouse.
+            oneOf:
+              - properties:
+                  policy:
+                    enum:
+                      - AntiAffinity
+                  manuallyManaged:
+                    not: { }
```
[the remark this answers is not in the document]
Comment (artbin) [reply ee/candi/cloud-providers/openstack/openapi/cluster_configuration.yaml]: done

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @artbin. I've started the workflow for label 'e2e/run/openstack'...

**Comment by artbin** [comment]
/e2e/abort openstack;Standard;containerd;1.25 8448284328 8448284328-1-con-1-25 /sys/deckhouse-oss/install:pr7312 debian@185.91.53.192

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @artbin. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @artbin. I've started the workflow for label 'e2e/run/openstack'...

**Review by nabokihms — APPROVED** [review]

**Pull request merged**